src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs

```diff
@@ -189,32 +280,61 @@ public void Dispose()
 
         internal AbstractProject GetProject(ProjectId id)
         {
-            AbstractProject project;
-            _projectMap.TryGetValue(id, out project);
-            return project;
+            lock (_gate)
+            {
+                AbstractProject project;
+                _projectMap.TryGetValue(id, out project);
+                return project;
+            }
         }
 
         /// <summary>
         /// Add a project to the workspace.
+        /// If invoked on the foreground thread, the add is executed right away.
+        /// Otherwise, the add is scheduled on foreground task scheduler.
```

Comment:
Why is this scheduled for later? Shouldn't we be pushing it to the workspace asynchronously and then dealing with just the open file handling on the UI thread?